File: Crio/crio_programming_interview_problems-master/StockSpan/StockSpan.java
import java.util.*;

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
class Pair {
    private Integer x;
    private Integer y;
    Pair(Integer x, Integer y) {
        this.x = x;
        this.y = y;
    }
    public Integer getX() {
        return x;
    }
    public Integer getY() {
        return y;
    }
}
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

class StockSpan {

	static int [] stockSpan(int ar[], int n) {
		// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
		Stack< Pair > s = new Stack< Pair >();

		int answer[] = new int[n];
		for (int i = 0 ; i < n; i++) {
			while (!s.isEmpty() && ((s.peek()).getX() <= ar[i]))
				s.pop();
			if (s.empty())
				answer[i] = i + 1;
			else
				answer[i] = (i + 1) - (s.peek()).getY();
			Pair p = new Pair(ar[i], i + 1);
			s.push(p);
		}
		return answer;
		// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int ar[] = new int[n];
		for (int i = 0; i < n; i++) {
			ar[i] = sc.nextInt();
		}
		int [] answer = stockSpan(ar, n);
		for (int i = 0; i < answer.length; i++)
			System.out.print(answer[i] + " ");
	}
}

File: Crio/crio_programming_interview_problems-master/StockSpan/StockSpan.cpp

#include <bits/stdc++.h>
using namespace std;
vector<int> stockSpan(vector<int> &A )
{
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    stack<int> s;
    int n = A.size();
    vector<int> ans(A.size());
    for ( int i = 0; i < n; i++) {
        while (!s.empty() && A[s.top()] <= A[i])
            s.pop();
        if (s.empty())
        {
            ans[i] = i + 1;
            s.push(i);
        } else {
            ans[i] = i - s.top();
            s.push(i);
        }
    }
    return ans;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}
int main()
{
    int n;
    cin >> n;
    vector<int> A(n);
    for ( int i = 0; i < n; i++)
        cin >> A[i];
    vector<int> ans = stockSpan(A);
    for ( int i = 0; i < n; i++)
        cout << ans[i] << " ";
    return 0;
}






